Hw03 generic finish

psannikov opened the pull request.

psannikov committed:
first commit hw03-generic

psannikov committed:
finish hw03-generic

FlameXander commented on hw03-generic/src/main/java/ru/otus/pro/psannikov/Box.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.Arrays;
+
+public class Box <T extends Fruit>{
+    ArrayList<T> fruits;
```
Почему поле не приватное? И обычно используют интерфейс List

psannikov replied:
исправил

FlameXander commented on hw03-generic/src/main/java/ru/otus/pro/psannikov/Box.java:
```diff
@@ -0,0 +1,54 @@
+        }
+        return this.weight() == otherBox.weight();
+    }
+    public void pour(Box <T> otherBox) {
```
Если otherBox == нулл или this, то будут проблемы

FlameXander commented on hw03-generic/src/main/java/ru/otus/pro/psannikov/Box.java:
```diff
@@ -0,0 +1,54 @@
+        return fruits;
+    }
+
+    public void addFruits(T...fruits) {
```
Реализация складывания фруктов в корректные коробки должна быть сделана на уровне дженериков, а не пачки условий

psannikov replied:
убрал проверки совместимости типов для коробки типа Fruit. Сейчас все не обощениях

FlameXander commented on hw03-generic/src/main/java/ru/otus/pro/psannikov/Box.java:
```diff
@@ -0,0 +1,54 @@
+package ru.otus.pro.psannikov;
```
В пакете еще было бы неплохо указать что это дз по дженерикам, или просто дз3

psannikov replied:
исправил

psannikov committed:
hw03-generic fix by comment pr:
1. private List fruit
2. check null and this in pour
3. replace if logic ing addFruit
4. rename package

FlameXander commented on hw03-generic/src/main/java/ru/otus/pro/psannikov/hw03_generic/Box.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public void pour(Box <T> otherBox) {
+        if (otherBox == null || this == otherBox) {} else {
```
Пустое тело if выглядит чуть-чуть странно)

```
if (otherBox == null || this == otherBox) {
  return;
}
for (T fruit : this.fruits) {
  otherBox.addFruits(fruit);
}
this.fruits.clear();
```

psannikov committed:
fix empty if

psannikov committed:
Merge branch 'master' into hw03-generic
